commit fc20d87173c4
Author: Manish Goregaokar <manishearth@gmail.com>
Date:   Tue Sep 26 23:56:15 2017 -0500

    Bug 1403282 - stylo: Don't error out on trailing whitespace in attr(). r=bz, a=sledru
    
    Source-Repo: https://github.com/servo/servo
    Source-Revision: 97227aaca3613e22afc832554d29f7f7fdfb7220
    
    --HG--
    extra : source : 4263e1c081c940b4eae7bcd388274ec8c5b5dd8b
---
 servo/components/style/values/specified/mod.rs | 44 ++++++++++++++------------
 1 file changed, 24 insertions(+), 20 deletions(-)

diff --git servo/components/style/values/specified/mod.rs servo/components/style/values/specified/mod.rs
index 856aecca90ac..72ed1b1de83a 100644
--- servo/components/style/values/specified/mod.rs
+++ servo/components/style/values/specified/mod.rs
@@ -739,28 +739,32 @@ impl Attr {
         let first = input.try(|i| i.expect_ident_cloned()).ok();
         if let Ok(token) = input.try(|i| i.next_including_whitespace().map(|t| t.clone())) {
             match token {
-                Token::Delim('|') => {}
+                Token::Delim('|') => {
+                    // must be followed by an ident
+                    let second_token = match *input.next_including_whitespace()? {
+                        Token::Ident(ref second) => second,
+                        ref t => return Err(BasicParseError::UnexpectedToken(t.clone()).into()),
+                    };
+
+                    let ns_with_id = if let Some(ns) = first {
+                        let ns = Namespace::from(ns.as_ref());
+                        let id: Result<_, ParseError> =
+                            get_id_for_namespace(&ns, context)
+                            .map_err(|()| StyleParseError::UnspecifiedError.into());
+                        Some((ns, id?))
+                    } else {
+                        None
+                    };
+                    return Ok(Attr {
+                        namespace: ns_with_id,
+                        attribute: second_token.as_ref().to_owned(),
+                    })
+                }
+                // In the case of attr(foobar    ) we don't want to error out
+                // because of the trailing whitespace
+                Token::WhiteSpace(_) => (),
                 ref t => return Err(BasicParseError::UnexpectedToken(t.clone()).into()),
             }
-            // must be followed by an ident
-            let second_token = match *input.next_including_whitespace()? {
-                Token::Ident(ref second) => second,
-                ref t => return Err(BasicParseError::UnexpectedToken(t.clone()).into()),
-            };
-
-            let ns_with_id = if let Some(ns) = first {
-                let ns = Namespace::from(ns.as_ref());
-                let id: Result<_, ParseError> =
-                    get_id_for_namespace(&ns, context)
-                    .map_err(|()| StyleParseError::UnspecifiedError.into());
-                Some((ns, id?))
-            } else {
-                None
-            };
-            return Ok(Attr {
-                namespace: ns_with_id,
-                attribute: second_token.as_ref().to_owned(),
-            })
         }
 
         if let Some(first) = first {

commit 20bbda2d12de
Author: Manish Goregaokar <manishearth@gmail.com>
Date:   Tue Sep 26 13:54:22 2017 -0700

    Bug 1403282 - stylo: Add reftests for whitespace in attr(). r=bz, a=sledru
    
    MozReview-Commit-ID: IMXFY7bpKGf
    
    --HG--
    extra : source : cfcb518088438415e5c6f7776dc268abd0b86119
---
 .../generated-content/attr-whitespace-ref.html     | 10 ++++
 .../generated-content/attr-whitespace.xhtml        | 60 ++++++++++++++++++++++
 layout/reftests/generated-content/reftest.list     |  1 +
 3 files changed, 71 insertions(+)

diff --git layout/reftests/generated-content/attr-whitespace-ref.html layout/reftests/generated-content/attr-whitespace-ref.html
new file mode 100644
index 000000000000..54a74fe4d81f
--- /dev/null
+++ layout/reftests/generated-content/attr-whitespace-ref.html
@@ -0,0 +1,10 @@
+<!DOCTYPE HTML>
+<html>
+<body>
+<div>HelloWorld</div>
+<div>HelloWorld</div>
+<div>HelloWorld</div>
+<div>Hello</div>
+<div>Hello</div>
+</body>
+</html>
diff --git layout/reftests/generated-content/attr-whitespace.xhtml layout/reftests/generated-content/attr-whitespace.xhtml
new file mode 100644
index 000000000000..e9b6554849b0
--- /dev/null
+++ layout/reftests/generated-content/attr-whitespace.xhtml
@@ -0,0 +1,60 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN"
+  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
+<html lang="en" xml:lang="en" xmlns="http://www.w3.org/1999/xhtml">
+<head>
+    <title>Whitespace in attributes</title>
+    <link rel="author" title="Manish Goregaokar" href="mailto:manishearth@gmail.com" />
+    <link rel="author" title="Mozilla Corporation" href="http://mozilla.com/" />
+    <link rel="help" href="https://drafts.csswg.org/css-values-3/#attr-notation"/>
+<style>
+@namespace ns url(http://www.example.com/ns);
+/* Spaces within attr() are fine*/
+#a::after {
+
+  content: attr(          
+      data-text  
+
+
+          );
+}
+#b::after {
+
+  content: attr(ns|data-text);
+}
+#c::after {
+
+  content: attr(          
+      ns|data-text  
+
+
+          );
+}
+
+/* No whitespace around the | */
+#d::after {
+
+  content: attr(          
+      ns      |data-text
+
+
+          );
+}
+#e::after {
+
+  content: attr(          
+      ns|       data-text
+
+
+          );
+}
+</style>
+</head>
+<body xmlns:ns="http://www.example.com/ns">
+<div id="a" data-text="World">Hello</div>
+<div id="b" ns:data-text="World">Hello</div>
+<div id="b" ns:data-text="World">Hello</div>
+<div id="d" ns:data-text="World">Hello</div>
+<div id="e" ns:data-text="World">Hello</div>
+</body>
+</html>
diff --git layout/reftests/generated-content/reftest.list layout/reftests/generated-content/reftest.list
index e07a0dd565a5..c789f1845421 100644
--- layout/reftests/generated-content/reftest.list
+++ layout/reftests/generated-content/reftest.list
@@ -19,3 +19,4 @@ fuzzy-if(OSX==1010,1,10) == table-parts-01.html table-parts-01-ref.html
 == before-style-sharing.html before-style-sharing-ref.html
 == transitive-style-invalidation.html transitive-style-invalidation-ref.html
 == dynamic-content.html dynamic-content-ref.html
+== attr-whitespace.xhtml attr-whitespace-ref.html
